R/derive_var_confirmation_flag.R

```diff
@@ -0,0 +1,329 @@
+#'   USUBJID AVISITN AVALC  AVAL AVISITN.join AVALC.join
+#'   <chr>     <dbl> <chr> <dbl>        <dbl> <chr>
+#'   1             1 Y         1            1 Y
+#'   1             1 Y         1            2 N
+#'   1             2 N         0            2 N
+#'   ```
+#'
+#'   ## Step 3
+#'
+#'   If `first_cond` is specified, for each observation of the input dataset the
+#'   joined dataset is restricted to observations up to the first observation
+#'   where `first_cond` is fulfilled (the observation fulfilling the condition
+#'   is included). If for an observation of the input dataset the condition is
+#'   not fulfilled, the observation is removed.
+#'
+#'   ## Step 4
+#'
+#'   The joined dataset is grouped by the observations from the input dataset
+#'   and restricted to the observations fulfilling the condition specified by
+#'   `filter`.
+#'
+#'   ## Step 5
+#'
+#'   The first observation of each group is selected
+#'
+#'   ## Step 6
+#'
+#'   The variable specified by `new_var` is added to the input dataset. It is
+#'   set to `true_value` for all observations which were selected in the
+#'   previous step. For the other observations it is set to `false_value`.
+#'
+#' @return The input dataset with the variable specified by `new_var` added.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords adam derivation
+#'
+#' @seealso [filter_confirmation()]
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#' library(admiral)
+#'
+#' # flag observations with a duration longer than 30 and
+#' # at, after, or up to 7 days before a COVID AE (ACOVFL == "Y")
+#' adae <- tribble(
+#'   ~USUBJID, ~ADY, ~ACOVFL, ~ADURN,
+#'   "1",        10, "N",          1,
+#'   "1",        21, "N",         50,
+#'   "1",        23, "Y",         14,
+#'   "1",        32, "N",         31,
+#'   "1",        42, "N",         20,
+#'   "2",        11, "Y",         13,
+#'   "2",        23, "N",          2,
+#'   "3",        13, "Y",         12,
+#'   "4",        14, "N",         32,
+#'   "4",        21, "N",         41
+#' )
+#'
+#' derive_var_confirmation_flag(
+#'   adae,
+#'   new_var = ALCOVFL,
+#'   by_vars = vars(USUBJID),
+#'   join_vars = vars(ACOVFL, ADY),
+#'   join_type = "all",
+#'   order = vars(ADY),
+#'   filter = ADURN > 30 & ACOVFL.join == "Y" & ADY >= ADY.join - 7
+#' )
+#'
+#' # flag observations with AVALC == "Y" and AVALC == "Y" at one subsequent visit
+#' data <- tribble(
+#'   ~USUBJID, ~AVISITN, ~AVALC,
+#'   "1",      1,        "Y",
+#'   "1",      2,        "N",
+#'   "1",      3,        "Y",
+#'   "1",      4,        "N",
+#'   "2",      1,        "Y",
+#'   "2",      2,        "N",
+#'   "3",      1,        "Y",
+#'   "4",      1,        "N",
+#'   "4",      2,        "N",
+#' )
+#'
+#' derive_var_confirmation_flag(
+#'   data,
+#'   by_vars = vars(USUBJID),
+#'   new_var = CONFFL,
+#'   join_vars = vars(AVALC, AVISITN),
+#'   order = vars(AVISITN),
+#'   filter = AVALC == "Y" & AVALC.join == "Y" & AVISITN < AVISITN.join
```

Comment:
looking through the examples again and trying to put myself in a neutral user perspective and i think the piece users will most struggle with understanding here is what's the difference when using `first_cond`, e.g. why would the above be different to:
```
first_cond = AVALC.join == "Y",
filter = AVALC == "Y" & AVISITN < AVISITN.join
```
Any way you think we could help explain this more or maybe via an example? i'm even struggling myself today to see the difference, but maybe it's just because it's first thing Monday morning...

Reply:
Good point! There is no difference. Both examples work. You need `first_cond` if `filter` uses summary functions which should be applied up to the confirmation assessment only, e.g., number of NEs between assessment and confirmation assessment should be at most one.

I have added this to the documentation of `first_cond`.

Reply:
I can see the benefit of having 2 input parameters ie. `filter` and `first_cond`, however I agree with @rossfarrugia it is quite confusing. Could the input parameters be renamed, so `filter` would always be used, and would not contain any summary functions (ie what `first_cond` is at the moment), and when summary functions are required you would use the second input parameter, called something like `filter_summary`?

Reply:
I think this could be misleading. I would not rename `filter` because this should be the main condition which defines whether an observation is flagged or not. I would rather rename `first_cond` because it seems that the purpose of the parameter is not clear. The purpose is to restrict the other observations for which `filter` is evaluated. (As an observation is flagged if `filter` is `TRUE` for any observation, the restriction affects the result only if summary functions are used in `filter`.)
Maybe we should rename `first_cond` to `max_obs_cond` or `last_obs_cond`?

What do you think?

Reply:
I see. I think test 2 is not a good example. For flagging CR assessments followed by a CR assessment I would rather use
```
derive_var_confirmation_flag(
      data,
      new_var = CONFFL,
      by_vars = vars(USUBJID),
      join_vars = vars(AVALC),
      join_type = "after",
      order = vars(AVISITN),
      filter = AVALC == "CR" & AVALC.join == "CR"
    )
```
I think this is clearer than the code in test 2. (We can not avoid that the code in test 2 works.)

Maybe we should remove test 2? The current code is testing the same as test 3 and the code above would test the same as test 1.